docs/decisions/0007-reducing-complexity.md

```diff
@@ -0,0 +1,39 @@
+
+### Reusable hooks
+
+Decision: Hooks intended to be reusable live in shared directories. Either the directory name or the filename they are in must signify that they are hooks.
+
+Reason: Easily distinguish hooks from other functions
+
+### Self-imports
+
+Decision: We avoid self-imports and replace them wherever we find them.
+
+Reason: See general principles (5). They have unnecessary complexity. The conceptual idea behind it is not so clear: the way we program in JS is usually a one-directional flow where we declare a const for example, either import it or declare it in the same file, then use it down the line and export it. Conceptually the idea that you declare a constant, export it, then import it into the same file, then export again makes everything pretty convoluted, and we also have extra syntactical overhead for it. The code itself does not explain why the logic is so complicated and what it is there for - because it is only there for the tests. From a code perspective it is not logical to use such a logic. We do not need self-imports for writing tests, other projects generally do not use them.
+
+Note: If you feel you really need something like this for your tests, consider this: instead of self-imports, exporting a class allows mocking of class methods, which also solves the problem and is already much better, since classes are a normal pattern and there is nothing bad about them. Still, this is not ideal because the code structure should not depend on the test. The best solution is to fix the problem you are having in the tests, not in the application code. There is a special (babel-plugin)[https://github.com/speedskater/babel-plugin-rewire/] for this which seems to solve the problem and which you should *only* add to the test environment.
+
+### component state
+
+Decision: We use useState in the way it is demonstrated in the React docs, and we do not export a constant `state` from our hooks.js files.
+
+Reason: See general principles (3, 5). React has developed an elegant, well-working and testable state hook, and we do not need to abstract that with unnecessary complexity. This is also not needed for testability. Mocking useState is very difficult because you are not intended to. If you really need to do so, you can still absolutely do it, and this is a testing issue and should not affect the code in any way. However, it is not at all recommended to test state directly. Testing state directly, e.g. by mocking useState, will lead to false positives and false negatives in the tests, and as of React 18 useState behavior is not deterministic in development (&test?) environments. There is also the possibility of working with `useReducer`, (React local state hook, not redux reducer), which should be a bit easier to test.
```

Comment:
Are you suggesting that configuring a test with a specific starting state ought to be done with the mutator returned by useState, and never by mocking useState itself? Or are you suggesting that a test should never be in the business of directly setting state as part of its setup? (This is a clarification question only).

Reply:
Yeah it's actually a whole topic and very interesting point, and there are many very interesting articles on the topic. I would love to talk about the way we test in React, but I think that is very out of scope here and a larger topic and more controversial, so I just added this note to explain why you can get around without needing this weird pattern. Generally the idea out there is that you don't test your React hooks and state directly because they are not pure functions. They are all about side effects that control the lifecycle of the React component. A state hook behaves differently in development, test and production environments starting from React 18, so it is not deterministic. The hook calling setState may work amazingly well but from calling it you cannot see what it does with the lifecycle of the component, so you often have false positives. You also may have false negatives because the lifecycle and state may be controlled by different elements that make it work nonetheless. Basically what we are testing are not state hooks, but the component state and lifecycle, and we need to make that behave like in reality. So what the recommended way of testing React state is is to test the whole component as your unit and test the behavior when the component renders, when it loads data, when a button is clicked and so on. So now what if you have a React hook that is built for many components to reuse? And we don't have many of those anyway. If the hook is just part of your one component, best to test it as part of that component. Otherwise, you would in theory build a mock/testing component and add the hook to that component; but react-testing-library has tools to do that for you so you don't have to worry about it actually.